MinaProtocol/mina
# Snapp test transaction tool

Explain your changes:
* Tool to generate  snapp transactions (dummy snapp) that update various fields. (also has #9860 fix)
* it prints the graphQL `send-snapp` that can be pasted on the graphql interface

Commands:
- [X] create-snapp-account
- [X] update app state
- [X] update permissions
- [X] upgrade-snapp
- [X] transfer-funds
- [X] update-snapp-uri
- [X] update-sequence-state
- [X] update-token-symbol
- and more

Explain how you tested your changes:
* create a snapp account and deploy a dummy snapp 
```
_build/default/src/app/snapp_test_transaction/snapp_test_transaction.exe create-snapp-account --fee-payer-key ../keys/wallet-keys/1  --receiver-amount 10 --snapp-account-key ../keys/wallet-keys/3 --nonce 0 
```
example output:
```
mutation MyMutation {
  __typename
  sendSnapp(input: {
        protocolState: {
      nextEpochData: {
        epochLength: null, 
        lockCheckpoint: null, 
        startCheckpoint: null, 
        seed: null, 
        ledger: {
          totalCurrency: null, 
          hash: null}}, 
      stakingEpochData: {
        epochLength: null, 
        lockCheckpoint: null, 
        startCheckpoint: null, 
        seed: null, 
        ledger: {
          totalCurrency: null, 
          hash: null}}, 
      globalSlotSinceGenesis: null, 
      globalSlotSinceHardFork: null, 
      totalCurrency: null, 
      minWindowDensity: null, 
      blockchainLength: null, 
      timestamp: null, 
      snarkedNextAvailableToken: null, 
      snarkedLedgerHash: null}, 
    otherParties: [ { 
        authorization: {  signature: "7mX39wTnDiiFrpttXR9nxnZR8SkJvPNrJmRRKkiVsWhUH7vGCvP2GXRZJpm5NAhiTMPjBQZnftd4id1zEYCBhYn5WabBtMf9" }, 
      data: {
        predicate: { nonce: "1" }, 
        body: {
          depth: 0, 
          callData: "0x0000000000000000000000000000000000000000000000000000000000000000", 
          sequenceEvents: [], 
          events: [], 
          delta: {sign: MINUS, magnitude: "10000000000"}, 
          tokenId: "1", 
          update: {
            timing: null, 
            tokenSymbol: null, 
            snappUri: null, 
            permissions: null, 
            verificationKey: null, 
            delegate: null, 
            appState: [null,null,null,null,null,null,null,null]}, 
          publicKey: "B62qiy32p8kAKnny8ZFwoMhYpBppM1DWVCqAPBYNcXnsAHhnfAAuXgg"}}
     }, { 
        authorization: {  signature: "7mXCnmWkAB9Z1V9yZ3Pp6Yqa1MgZoxYGPyCQ2tLs7VdUfBmxrWad1bpYmaskzPCXEgM9moXSkgUou2GH7C7HW9tvKBB4CKeG" }, 
      data: {
        predicate: {account:null, nonce:null}, 
        body: {
          depth: 0, 
          callData: "0x0000000000000000000000000000000000000000000000000000000000000000", 
          sequenceEvents: [], 
          events: [], 
          delta: {sign: PLUS, magnitude: "10000000000"}, 
          tokenId: "1", 
          update: {
            timing: null, 
            tokenSymbol: null, 
            snappUri: null, 
            permissions: 
      { stake: true
  , editState: Proof
  , send: Signature
  , receive: None
  , setDelegate: Signature
  , setPermissions: Signature
  , setVerificationKey: Signature
  , setSnappUri: Signature
  , editSequenceState: Proof
  , setTokenSymbol: Signature
      }
    , 
            verificationKey: 
          {verificationKey: "Q758dSnBCFdoR7jMXWsVxJxehfwXaaDuCifWh4WcGKJ6DpohD8QaxB4g3nFy6FspAqWxYPawMxtfUTPeFxvdSkh28hwrjMfKwZo7ZYtjVSqnPHaDGdgqfk8qiDKunsC4qxEXars48kQFKVCjavYqdjHWfmGcQ1RUKq58T7APomHLqXGJFm3TZh2mrhBXVd87txbjXfKaGqUquWrBvKYch3EdtTsSsKgBMheYyrcvQQ4Lh62Zs4FhsbcpYo6t37NVNVw2XB7hQczdym7Lg3vcD3wvPs33Ct4KN3n7JvZxNB5jHkDgKvhZyNGwd2sPJ4u6JEXEe7zbAJGyYUbAUYKJG1GEpzv55E6XBr8Vnu7WgdcVzZsqoMEKjmabSAT5ZNV9bp9eat5Qwa6BrEdyqXiK1mEL13fKxt8uCR3G6hwKWk7DXvHmmSbqMMARAf9Q2mWM1ceW88aJnpHmtbpqAiUaCApYUWRx6Z3vLcUQ42yk2m9MhgqXmXruZN5eM2kWJTR9FKdtKBqkqMTfeMonnTt4q2f97HZJdC7sADZLbAco81QJdiaKPnNagWnckcJsj6GUKtnQjmuHDKBSA4PuyAveiz4JHec57o8W2geEPQZFEKyBP5np1Uu241pZaCBP3C1yuaKHjwC5Wvw7HYC1YaGoaW2PLjvbPXMpYT2B1cLNs8HwGPiBXwhXfU8Shpz14irxnbG5sihdGu6doKhtZHk8z5iq9upyUph3VLeM4HbVq5JkCfWFCEeboq5Q11T6itfwCAcWWM9N3choMx6gHGFvSS7a7oCWkSuw924QqKUrmA1sPrdWPvhrXhSWqy3WiZFn5viNbojPikqSjjX1njFFq2qbtfJz7oP79TjkUMcS5rNsCPxyV5Tps54tsgB6hQPNiAGvtQ9xAnm8g6oNQuGozKqBmG2ovg1KwArERFM3pFQ5f7KFbgxiG9vFvhaHhsiNHCZwEhjQZjeQ1srhQyufkXeFztGUZvj2EbC85JmgLm251GUW4BeqcJt4LuJFNvUfjcabnSKLVeiiaBBiCaGt68vHW7Bybm7MCvKyyMVpwNsdCDvxs9NPznpZp7cxoXvkFeVVzjwh11iWMdFAax2uYtHybBQaVUBHahzkp2Dmg82Ci5XGwKTWFLpaJ8TPyjnvCnYz4iu8qzCQ8GDD9bxqGyMdb3pAKNWKp4vGAysTHFaDwczG6iegVMdPfuPkHazag3Cxk4NT58u66GD3P5R45H2swSisViXQ3UDNiCVpqFXpJutG97pA2mpZu47tYvNyajh6T3m1VkN3SbJzmZNnPtxksrkb3kBRssiFm2EWeYy5Ny7v8CQWYV6Qu5fH1UXesmxFBK7jSfz9adZ7rJnRe3ztndikviFwUy8DFW9R3FG44m5eZtpQeDWT8u9XMVcCdaU1piF5B3iBj4hkHHRgBrpciiZ4tY3WGAnPQRRFdCgwcHjxNqZvAdZ1gNA159RVqgphH8JmshCBjtSBBHcgsLZ1LytyniJ6TigXzQSRcN2dnt1A3QzCPTw2tdkUPkxTzuDW3GWyYhfgoF7L9amm28JEHHdvvfXGf8fvUj6318MfEQur1MKdFTREfCh3FBiapk1WRyPPHEvwfEgVGJgfdW18ANbuGKetm", hash: "0xa6014879fa95df23d9b72e8b654f054791ebd4eb05d0c62e5b4dd761b9c57705"}
        , 
            delegate: null, 
            appState: [null,null,null,null,null,null,null,null]}, 
          publicKey: "B62qrbENDJv8qrQ5sJoGhzmA7rxT571StwTzqZ8veiPMh5guLcPQ6XW"}}
     }], 
    feePayer: {
        authorization: "7mXBBnDdn1cfU6VqbhHYxQp4BYvxvU6N2PNoFiqdFYyHHxvBPDecjgGXeHZY95L8JuaB7PwmgpLczhQHCdkJTCATtUGdEoeH", 
      data: {
        predicate: "0", 
        body: {
          depth: "0", 
          callData: "0x0000000000000000000000000000000000000000000000000000000000000000", 
          sequenceEvents:[], 
          events: [], 
          fee: "1000000000", 
          update: {
            timing: null, 
            tokenSymbol: null, 
            snappUri: null, 
            permissions: null, 
            verificationKey: null, 
            delegate: null, 
            appState: [null,null,null,null,null,null,null,null]}, 
          pk: "B62qiy32p8kAKnny8ZFwoMhYpBppM1DWVCqAPBYNcXnsAHhnfAAuXgg" }
        } }})
}

```
* update app state

```
 _build/default/src/app/snapp_test_transaction/snapp_test_transaction.exe update-state --fee-payer-key ../keys/wallet-keys/1  --nonce 0 --snapp-account-key ../keys/wallet-keys/3 --snapp-state 1 --snapp-state 2 --snapp-state 3 --snapp-state 4 --snapp-state 5 --snapp-state 6 --snapp-state 7 --snapp-state 8

```


*  update permissions
```
_build/default/src/app/snapp_testnsaction/snapp_test_transaction.exe update-permissions --fee-payer-key ../keys/wallet-keys/1  --nonce 0 --snapp-account-key ../keys/wallet-keys/3 --current-auth Signature --edit-stake Proof --receive None --set-permissions Proof --set-delegate Signature --set-verification-key Proof --set-snapp-uri None --set-sequence-state Proof --set-token-symbol Signature --send Signature
```




Checklist:

- [ ] Modified the current draft of release notes with details on what is completed or incomplete within this project
  - Mention expected invariants, implicit constraints
  - Document test purpose, significance of failures
  - Test names should reflect their purpose
- [ ] Does this close issues? List them

* Closes #9834
